tests/Unit/DataStructures/Tensor/Expressions/Test_Contract.cpp

```diff
@@ -652,11 +652,136 @@ void test_contractions_rank4(const DataType& used_for_size) noexcept {
   CHECK(RjIiJ_contracted.get() == expected_RjIiJ_sum);
 }
 
+template <typename DataType>
+void test_time_index(const DataType& used_for_size) noexcept {
+  Tensor<DataType, Symmetry<4, 3, 2, 1>,
+         index_list<SpacetimeIndex<3, UpLo::Up, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Up, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>,
+                    SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>>>
+      R(used_for_size);
+  create_tensor(make_not_null(&R));
+  // Contract RHS tensor with time index to a LHS tensor without a time index
+  // Use explicit type (vs auto) for LHS Tensor so the compiler checks the
+  // return type of `evaluate`
+  // \f$L_{b} = R^{at}{}_{ab}\f$
+  const Tensor<DataType, Symmetry<1>,
+               index_list<SpacetimeIndex<3, UpLo::Lo, Frame::Inertial>>>
+      R_contracted_1 =
+          TensorExpressions::evaluate<ti_b>(R(ti_A, ti_T, ti_a, ti_b));
+
+  for (size_t b = 0; b < 4; b++) {
+    DataType expected_R_sum_1 = make_with_value<DataType>(used_for_size, 0.0);
+    for (size_t a = 0; a < 4; a++) {
+      expected_R_sum_1 += R.get(a, 0, a, b);
+    }
+    CHECK_ITERABLE_APPROX(R_contracted_1.get(b), expected_R_sum_1);
+  }
+
+  // Contract RHS tensor with upper and lower time index to a LHS tensor without
+  // time indices
+  // \f$L = R^{at}{}_{at}\f$
+  //
+  // Makes sure that `TensorContract` does not get confused by the presence of
+  // an upper and lower time index in the RHS tensor, which is different than
+  // the presence of an upper and lower generic index
+  const Tensor<DataType> R_contracted_2 =
+      TensorExpressions::evaluate(R(ti_A, ti_T, ti_a, ti_t));
```

Comment:
(chosen a line at random) What do you think about naming the tensor indices `ti::a`, `ti::A`, etc., i.e. replacing the underscore with a namespace? I think that would improve readability and user-friendliness. If you agree, then I would suggest making that change in a follow-up PR (not in this one!) relatively soon, before you tensor expressions start to get uses in too many places (like your work on the CCZ4 system). It should be a relativity straight-forward regex-replace.

Reply:
I don't see any reason to put in this work. I personally prefer the current syntax

Reply:
I don't have a strong preference either way. Since it seems there is disagreement, I'm happy to ask others for input, if you two think that's a good idea.

Reply:
Let's defer this, nothing to do in this PR